# King County Housing Linear Regression Project


* Student name: Seyma Tas
* Student pace:  Full time
* Scheduled project review date/time: 05-04-2020
* Instructor name: Amber Yandow
* Blog post URL:


## OSEMN Approach
1 - Obtaining the data 2 - Scrubbing the data 3 - Exploring the data 4 - Modeling the data 5 - iNterpreting the results


![title](picture2.png)


## Questions:

1) What are the main factors to predict the price of a house?

2) How much does a middle income family should invest to buy a standard house?
   
   or What are the properties of a house you can buy in King County if your budget is $400,000???
    
3) How can a middle income family increase the price of their house by %20 by spending %5 of the house cost???

4) Basement
 

In [1]:
%autosave 10

Autosaving every 10 seconds


The following cell is to print all the interactive output without resorting to print. 

In [2]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Import the necessary packages

First, we’ll import all of our necessary packages.

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

from scipy import stats
import scipy.stats as stats
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.outliers_influence import variance_inflation_factor

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score

sns.set(style='darkgrid')

Pandas has an options system that lets you customize some aspects of its behaviour, display-related options being those the user is most likely to adjust. In the following cell we are setting options to display more raws and columns.

In [ ]:
# ## Display more information 

# pd.set_option('display.max_rows', None)
# pd.set_option('display.max_columns', None)
# pd.set_option('display.width', None)
# pd.set_option('display.max_colwidth', -1)

The following code is to hide the warning boxes.

In [4]:
import warnings
warnings.filterwarnings('ignore')

## Import the data set

Next, we’ll import our data and see what we’re working with. 

In [18]:
df=pd.read_csv('kc_house_data.csv')

Lets see what the data looks like by functions head(), tail(), shape(), info() and describe().

In [19]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0         NaN   0.0  ...      7        1180            0.0   
1      7242     2.0         0.0   0.0  ...      7        2170          400.0   
2     10000     1.0         0.0   0.0  ...      6         770            0.0   
3      5000     1.0         0.0   0.0  ...      7        1050          910.0   
4      8080     1.0         0.0   0.0  ...      8        1680            0.0   

  yr_built  yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0     1955           0.0    98178  47.5112 -122.257           1340        5650  
1     1951        1991.0    98125  47.7210 -122.319           1690        7639  
2     1933           NaN    98028  47.7379 -122.233           2720        8062  
3     1965           0.0    98136  47.5208 -122.393           1360        5000  
4     1987           0.0    98074  47.6168 -122.045           1800        7503  

[5 rows x 21 columns]

## Information about the data

### Column names and descriptions

* **id** - unique identified for a house
* **date** - Date house was sold
* **price** - Price is prediction target
* **bedrooms** - Number of Bedrooms/House
* **bathrooms** - Number of bathrooms/bedrooms
* **sqft_living** - Square footage of the home
* **sqft_lot** - Square footage of the lot
* **floors** - Total floors (levels) in house
* **waterfront** - House which has a view to a waterfront
* **view** - If the house has a view or not
* **condition** - How good the condition is ( Overall )
* **grade** - overall grade given to the housing unit, based on King County grading system
* **sqft_above** - square footage of house apart from basement
* **sqft_basement** - square footage of the basement
* **yr_built** - Built Year
* **yr_renovated** - Year when house was renovated
* **zipcode** - zipcode of the house
* **lat** - Latitude coordinate
* **long** - Longitude coordinate
* **sqft_living15** - The square footage of interior housing living space for the nearest 15 neighbors
* **sqft_lot15** - The square footage of the land lots of the nearest 15 neighbors


### Condition

More detailed information about condition from the [King County's website](https://blue.kingcounty.com/Assessor/eRealProperty/ResidentialGlossary.aspx?idx=viewall&Parcel=7960900070&AreaReport=http://www.KingCounty.gov/depts/Assessor/Reports/area-reports/2019/residential-northeast/033.aspx#BuildingGrade).

(Relative to age and grade. Coded 1-5.)

1 = Poor- Worn out. Repair and overhaul needed on painted surfaces, roofing, plumbing, heating and numerous functional inadequacies. Excessive deferred maintenance and abuse, limited value-in-use, approaching abandonment or major reconstruction; reuse or change in occupancy is imminent. Effective age is near the end of the scale regardless of the actual chronological age.

2 = Fair- Badly worn. Much repair needed. Many items need refinishing or overhauling, deferred maintenance obvious, inadequate building utility and systems all shortening the life expectancy and increasing the effective age.

3 = Average- Some evidence of deferred maintenance and normal obsolescence with age in that a few minor repairs are needed, along with some refinishing. All major components still functional and contributing toward an extended life expectancy. Effective age and utility is standard for like properties of its class and usage.

4 = Good- No obvious maintenance required but neither is everything new. Appearance and utility are above the standard and the overall effective age will be lower than the typical property.

5= Very Good- All items well maintained, many having been overhauled and repaired as they have shown signs of wear, increasing the life expectancy and lowering the effective age with little deterioration or obsolescence evident with a high degree of utility.




### Building Grade

More detailed information about grade from the [King County's website](https://blue.kingcounty.com/Assessor/eRealProperty/ResidentialGlossary.aspx?idx=viewall&Parcel=7960900070&AreaReport=http://www.KingCounty.gov/depts/Assessor/Reports/area-reports/2019/residential-northeast/033.aspx#BuildingGrade).

Represents the construction quality of improvements. Grades run from grade 1 to 13. Generally defined as:

1-3	= Falls short of minimum building standards. Normally cabin or inferior structure

4	= Generally older, low quality construction. Does not meet code.

5	= Low construction costs and workmanship. Small, simple design.

6	= Lowest grade currently meeting building code. Low quality materials and simple designs.

7	= Average grade of construction and design. Commonly seen in plats and older sub-divisions.

8	= Just above average in construction and design. Usually better materials in both the exterior and interior finish work.

9	= Better architectural design with extra interior and exterior design and quality.

10	= Homes of this quality generally have high quality features. Finish work is better and more design quality is seen in the floor plans. Generally have a larger square footage.

11	= Custom design and higher quality finish work with added amenities of solid woods, bathroom fixtures and more luxurious options.

12	= Custom design and excellent builders. All materials are of the highest quality and all conveniences are present.

13	= Generally custom designed and built. Mansion level. Large amount of highest quality cabinet work, wood trim, marble, entry ways etc.


### Drop unnecessary columns

Id column doesn't tell us anything so we drop this column. 
Date column tells us the date the house was sold. All the dates are between 2014-05-02 and 2015-05-27. So we don't need to calculate inflation. But the sale month tells us information about sale price. May must be the month many houses are sold and the prices must be higher in May. Lets take the month and drop the date.

In [20]:
df['date'] = pd.to_datetime(df['date'])

In [21]:
df['date'].min()

Timestamp('2014-05-02 00:00:00')

In [22]:
df['date'].max()

Timestamp('2015-05-27 00:00:00')

In [23]:
df['sale_month'] = df['date'].map(lambda x: x.month)

In [24]:
df=df.drop(['id','date'],axis=1)

# Scraping and cleaning data

In [11]:
df.shape

(21597, 20)

We have 20 columns and 21597 rows,our sample size is enough to predict the price. There won't be problems in test and train splits. 

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 20 columns):
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
sale_month       21597 non-null int64
dtypes: float64(8), int64(11), object(1)
memory usage: 3.3+ MB


According to the information of data; we have some missing values in waterfront, view and year_renovated columns. We need to deal with missing data. Data is clean. It is necessary to inspect sqft_basement column. It should be an integer not a string.

### Numerical Data Stored as Strings

In [13]:
df.sqft_basement.value_counts()

0.0       12826
?           454
600.0       217
500.0       209
700.0       208
          ...  
2490.0        1
3000.0        1
2250.0        1
1008.0        1
2500.0        1
Name: sqft_basement, Length: 304, dtype: int64

sqft_basement column is string because of 454 '?'. First we replace '?' with '0'. Then we change the data type to float and we find the 'sqft_basement' column from 'sqft_living' and 'sqft_above' columns.

In [14]:
# Replace '?' with '0'.
df['sqft_basement']=df['sqft_basement'].map(lambda x:float(x.replace('?', '0')))


In [15]:
#Replace 0 with difference between sqft_above and sqft_living
df['sqft_basement'].replace(to_replace='0', value = abs(df['sqft_living']-df['sqft_above']), inplace=True)


There is a house with 33 bedrooms. But its price is $640,000. Number of bathrooms is 1.75. It is a 3 bedroom house. 

In [27]:
df[df.bedrooms==33]

Empty DataFrame
Columns: [price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15, sale_month]
Index: []

In [26]:
# Change 33 bedrooms to 3 bedrooms.
df['bedrooms'].replace(to_replace=33, value =3, inplace=True)


In [28]:
df.describe().round(1)

price  bedrooms  bathrooms  sqft_living   sqft_lot   floors  \
count    21597.0   21597.0    21597.0      21597.0    21597.0  21597.0   
mean    540296.6       3.4        2.1       2080.3    15099.4      1.5   
std     367368.1       0.9        0.8        918.1    41412.6      0.5   
min      78000.0       1.0        0.5        370.0      520.0      1.0   
25%     322000.0       3.0        1.8       1430.0     5040.0      1.0   
50%     450000.0       3.0        2.2       1910.0     7618.0      1.5   
75%     645000.0       4.0        2.5       2550.0    10685.0      2.0   
max    7700000.0      11.0        8.0      13540.0  1651359.0      3.5   

       waterfront     view  condition    grade  sqft_above  yr_built  \
count     19221.0  21534.0    21597.0  21597.0     21597.0   21597.0   
mean          0.0      0.2        3.4      7.7      1788.6    1971.0   
std           0.1      0.8        0.7      1.2       827.8      29.4   
min           0.0      0.0        1.0      3.0       370.0    1900.0   
25%           0.0      0.0        3.0      7.0      1190.0    1951.0   
50%           0.0      0.0        3.0      7.0      1560.0    1975.0   
75%           0.0      0.0        4.0      8.0      2210.0    1997.0   
max           1.0      4.0        5.0     13.0      9410.0    2015.0   

       yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  \
count       17755.0  21597.0  21597.0  21597.0        21597.0     21597.0   
mean           83.6  98078.0     47.6   -122.2         1986.6     12758.3   
std           399.9     53.5      0.1      0.1          685.2     27274.4   
min             0.0  98001.0     47.2   -122.5          399.0       651.0   
25%             0.0  98033.0     47.5   -122.3         1490.0      5100.0   
50%             0.0  98065.0     47.6   -122.2         1840.0      7620.0   
75%             0.0  98118.0     47.7   -122.1         2360.0     10083.0   
max          2015.0  98199.0     47.8   -121.3         6210.0    871200.0   

       sale_month  
count     21597.0  
mean          6.6  
std           3.1  
min           1.0  
25%           4.0  
50%           6.0  
75%           9.0  
max          12.0

Most houses were build around 1975; sold for $450K; had square footage of living spaces around 1910. You can also note, that distributions of price and variables such as sqft_lot are skewed to the upper tail.

## Detecting and Dealing With Null Values

In [29]:
print(df.isna().sum())

price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
sale_month          0
dtype: int64


We need to deal with missing values in waterfront, view and year_renovated columns.

#### Waterfront and view columns

In [34]:
df['view'].value_counts()

0.0    19422
2.0      957
3.0      508
1.0      330
4.0      317
Name: view, dtype: int64

Most of the houses don't have view. Houses in Seattle might have both ocean, lake and mountain views.

In [35]:
len(df[(df.view==4)&(df.waterfront>0)])

123

Almost %38 of the houses with good views are waterfront homes. The others must have Mount Rainier view or downtown view.

We are trying to decide what to do with missing values in waterfront column.

In [ ]:
df['view'].isna().sum()

In [ ]:
feats = ['view', 'waterfront']
df[feats].corr()

In [ ]:
print('Number missing both:',
      len(df[(df.view.isnull())
      & (df.waterfront.isnull())]))

In [33]:
len(df[(df.view==0)&(df.waterfront>0)])

0

In [ ]:
len(df[(df.view==2)&(df.waterfront.isna())])

In [ ]:
len(df[(df.view==3)&(df.waterfront>0)])

Waterfront and view columns are corrolated. If there was a view or waterfront is would going to be written. We can say missinf values in waterfront column can be filled with '0'.

In [ ]:
df['view'].fillna(0,inplace=True)

In [ ]:
df['waterfront'].value_counts()

In [ ]:
df['waterfront'].nunique()

In [ ]:
df['waterfront'].isna().sum()

In [ ]:
df['waterfront'].fillna(0,inplace=True)

In [ ]:
feats = ['view', 'waterfront']
df[feats].corr()

#### Year_renovated column

In [ ]:
df.yr_renovated.fillna(0, inplace=True)

## Check the distributions

In [ ]:
df.hist(figsize=(20,15))
plt.show()


### Outliers

In [ ]:
Q1 = df.quantile(0.001)
Q3 = df.quantile(0.999)
IQR = Q3 - Q1

df = df[~((df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))).any(axis=1)]

## Log transformation of data

In [ ]:
non_normal = ['price','sqft_living', 'sqft_lot','sqft_above','sqft_basement',
              'yr_built', 'yr_renovated','sqft_living15', 'sqft_lot15']
for feat in non_normal:
    df[feat] = df[feat].map(lambda x: np.log1p(x))
# pd.plotting.scatter_matrix(df[x_cols], figsize=(10,12));

In [ ]:
df.head()

## Normalizing the data

In [ ]:
def norm_feat(series):
    return (series - series.mean())/series.std()
for feat in ['bathrooms', 'sqft_living', 'sqft_lot','sqft_above', 'sqft_basement', 'yr_built','yr_renovated','sqft_living15', 'sqft_lot15']:
    df[feat] = norm_feat(df[feat])


In [ ]:
df.describe().round(2)

In [ ]:
df.to_csv("king_county_dataset_cleaned.csv", index=False)


## Check the distributions again!

In [ ]:
df.hist(figsize=(20,15))
plt.show()

## One-Hot Encoding Categorical Columns


#### Identifying Categorical Variables

These columns can be thought as categorical values. Bedrooms, floors, waterfront, wiev, grade, zipcode and year renovated.
Let's explore them one by one.

In [ ]:
df['bedrooms'] = df['bedrooms'].astype('category')
df['floors'] = df['floors'].astype('category')
df['waterfront'] = df['waterfront'].astype('category')
df['view'] = df['view'].astype('category')
df['condition'] = df['condition'].astype('category')
df['grade'] = df['grade'].astype('category')
df['zipcode'] = df['zipcode'].astype('category')

In [ ]:
df.columns

In [ ]:
feats = ['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
       'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'sqft_living15', 'sqft_lot15']
df_feats = df[feats]
df_feats_zipcode=df_feats.copy()
df_feats = pd.get_dummies(df_feats, drop_first=True)

In [ ]:
df_feats.columns

In [ ]:
df_feats.head()

In [ ]:
df_feats.columns

# Exploring the data

In [ ]:
df.head()

In [ ]:
df_feats.head()

In [ ]:
df_plot=df[['price', 'bedrooms','bathrooms','sqft_living', 'sqft_lot', 'sqft_above', 'sqft_basement', 'yr_built', 'zipcode','lat','long','sqft_living15','sqft_lot15']].copy()

In [ ]:
sns.pairplot(df_plot)
plt.show()

## Check for Linearity

jointplots

## Check for Multicollinearity

In [ ]:
# df_target = df_no_outliers.loc['price']
df_heatmap=df.drop(['price','zipcode', 'lat', 'long'],axis=1)
plt.figure(figsize=(20,20))
ax=sns.heatmap(df_heatmap.corr(), center=0, linewidths=.5,annot=True);
bottom, top = ax.get_ylim()
ax.set_ylim(bottom + 0.5, top - 0.5)
plt.show()

In [ ]:
df_heatmap.corr() > 0.75


# Modeling the data

## Baseline Model 

In [ ]:
outcome = 'price'
x_cols = [ 'bathrooms', 'sqft_living', 'sqft_lot', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'sqft_living15',
       'sqft_lot15', 'bedrooms_2', 'bedrooms_3', 'bedrooms_4', 'bedrooms_5',
       'bedrooms_6', 'bedrooms_7', 'bedrooms_8', 'bedrooms_9', 'bedrooms_10',
       'bedrooms_11', 'floors_1_5', 'floors_2_0', 'floors_2_5', 'floors_3_0',
       'floors_3_5', 'waterfront_1_0', 'view_1_0', 'view_2_0', 'view_3_0',
       'view_4_0', 'condition_2', 'condition_3', 'condition_4', 'condition_5',
       'grade_4', 'grade_5', 'grade_6', 'grade_7', 'grade_8', 'grade_9',
       'grade_10', 'grade_11', 'grade_12', 'grade_13', 'zipcode_98002',
       'zipcode_98003', 'zipcode_98004', 'zipcode_98005', 'zipcode_98006',
       'zipcode_98007', 'zipcode_98008', 'zipcode_98010', 'zipcode_98011', 
       'zipcode_98019', 'zipcode_98022', 'zipcode_98023', 'zipcode_98024',
       'zipcode_98027', 'zipcode_98028', 'zipcode_98029', 'zipcode_98030',
       'zipcode_98031', 'zipcode_98032', 'zipcode_98033', 'zipcode_98034',
       'zipcode_98038', 'zipcode_98039', 'zipcode_98040', 'zipcode_98042',
       'zipcode_98045', 'zipcode_98052', 'zipcode_98053', 'zipcode_98055',
       'zipcode_98056', 'zipcode_98058', 'zipcode_98059', 'zipcode_98065',
       'zipcode_98070', 'zipcode_98072', 'zipcode_98074', 'zipcode_98075',
       'zipcode_98077', 'zipcode_98092', 'zipcode_98102', 'zipcode_98103',
       'zipcode_98105', 'zipcode_98106', 'zipcode_98107', 'zipcode_98108',
       'zipcode_98109', 'zipcode_98112', 'zipcode_98115', 'zipcode_98116',
       'zipcode_98117', 'zipcode_98118', 'zipcode_98119', 'zipcode_98122',
       'zipcode_98125', 'zipcode_98126', 'zipcode_98133', 'zipcode_98136',
       'zipcode_98144', 'zipcode_98146', 'zipcode_98148', 'zipcode_98155',
       'zipcode_98166', 'zipcode_98168', 'zipcode_98177', 'zipcode_98178',
       'zipcode_98188', 'zipcode_98198', 'zipcode_98199','zipcode_98014']


In [ ]:
# df_feats.columns = [col.replace('.', '_', inplace=True) for col in df_feats.columns]
# df_feats.columns = [col.replace(' ', '_', inplace=True) for col in df_feats.columns]
df_feats.columns = df_feats.columns.str.replace(r"[.]", "_")

In [ ]:
df_feats.columns[:50]

In [ ]:
df_feats.columns[50:]

In [ ]:
# Fitting the actual model
predictors = '+'.join(x_cols)
formula = outcome + '~' + predictors
model = ols(formula=formula, data=df_feats).fit()
model.summary()

In [ ]:
fig = sm.graphics.qqplot(model.resid, dist=stats.norm, line='45', fit=True)

In [ ]:
plt.scatter(model.predict(df_feats[x_cols]), model.resid)
plt.plot(model.predict(df_feats[x_cols]), [0 for i in range(len(df_feats))])
plt.show()

In [ ]:
X = df_feats[x_cols]

In [ ]:
vif = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
list(zip(x_cols, vif))

## K-fold Cross Validation for Model 1

In [ ]:
X = df_feats[x_cols]
y = df_feats['price']

linreg = LinearRegression()
cvs_10=cross_val_score(linreg, X, y, cv=10)
cvs_10_res_mean = round(np.mean(cvs_10),5)
cv_10_res_std = round(np.std(cvs_10),5)
print(f'Mean cross validation score for 10 fold is {cvs_10_res_mean}')
print(f'Standard deviation of the score is {cv_10_res_std}')


## Perform stepwise selection

In [ ]:
import statsmodels.api as sm

def stepwise_selection(X, y, 
                       initial_list=[], 
                       threshold_in=0.01, 
                       threshold_out = 0.05, 
                       verbose=True):
    """ Perform a forward-backward feature selection 
    based on p-value from statsmodels.api.OLS
    Arguments:
        X - pandas.DataFrame with candidate features
        y - list-like with the target
        initial_list - list of features to start with (column names of X)
        threshold_in - include a feature if its p-value < threshold_in
        threshold_out - exclude a feature if its p-value > threshold_out
        verbose - whether to print the sequence of inclusions and exclusions
    Returns: list of selected features 
    Always set threshold_in < threshold_out to avoid infinite looping.
    See https://en.wikipedia.org/wiki/Stepwise_regression for the details
    """
    included = list(initial_list)
    while True:
        changed=False
        # forward step
        excluded = list(set(X.columns)-set(included))
        new_pval = pd.Series(index=excluded)
        for new_column in excluded:
            model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit()
            new_pval[new_column] = model.pvalues[new_column]
        best_pval = new_pval.min()
        if best_pval < threshold_in:
            best_feature = new_pval.idxmin()
            included.append(best_feature)
            changed=True
            if verbose:
                print('Add  {:30} with p-value {:.6}'.format(best_feature, best_pval))

        # backward step
        model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included]))).fit()
        # use all coefs except intercept
        pvalues = model.pvalues.iloc[1:]
        worst_pval = pvalues.max() # null if pvalues is empty
        if worst_pval > threshold_out:
            changed=True
            worst_feature = pvalues.argmax()
            included.remove(worst_feature)
            if verbose:
                print('Drop {:30} with p-value {:.6}'.format(worst_feature, worst_pval))
        if not changed:
            break
    return included

In [ ]:
X = df_feats[x_cols]
# X = df_feats.drop('price', axis=1)
y = df_feats['price']

result = stepwise_selection(X, y, verbose = False)
print('resulting features:')
print(result)

In [ ]:
a=['grade_6', 'sqft_living', 'grade_7', 'grade_8', 'grade_5', 'zipcode_98004', 'zipcode_98023', 'grade_9', 'waterfront_1_0', 'sqft_living15', 'zipcode_98003', 'zipcode_98092', 'zipcode_98038', 'zipcode_98002', 'zipcode_98032', 'zipcode_98039', 'bathrooms', 'zipcode_98033', 'zipcode_98040', 'zipcode_98112', 'zipcode_98115', 'zipcode_98117', 'zipcode_98103', 'zipcode_98199', 'zipcode_98105', 'zipcode_98119', 'zipcode_98107', 'zipcode_98116', 'zipcode_98109', 'zipcode_98122', 'zipcode_98102', 'view_4_0', 'zipcode_98006', 'zipcode_98052', 'zipcode_98136', 'zipcode_98144', 'sqft_lot', 'zipcode_98008', 'zipcode_98005', 'grade_4', 'zipcode_98053', 'view_3_0', 'zipcode_98029', 'zipcode_98126', 'zipcode_98125', 'zipcode_98034', 'zipcode_98177', 'zipcode_98007', 'condition_5', 'zipcode_98074', 'zipcode_98075', 'grade_10', 'zipcode_98027', 'view_2_0', 'sqft_above', 'zipcode_98133', 'zipcode_98118', 'zipcode_98072', 'zipcode_98155', 'condition_4', 'condition_3', 'view_1_0', 'zipcode_98011', 'zipcode_98065', 'zipcode_98028', 'zipcode_98077', 'zipcode_98106', 'yr_renovated', 'zipcode_98108', 'zipcode_98059', 'zipcode_98056', 'zipcode_98024', 'grade_11', 'floors_3_0', 'zipcode_98166', 'zipcode_98146', 'zipcode_98045', 'zipcode_98019', 'zipcode_98014', 'zipcode_98070', 'zipcode_98010', 'zipcode_98058', 'zipcode_98178', 'zipcode_98055', 'sqft_basement', 'sqft_lot15', 'bedrooms_7', 'zipcode_98148', 'grade_13', 'bedrooms_2', 'condition_2', 'floors_2_0', 'bedrooms_6', 'zipcode_98188', 'zipcode_98168', 'bedrooms_5', 'bedrooms_9']

In [ ]:
len(a)

In [ ]:
b=['grade_7', 'sqft_living', 'zipcode_98004', 'waterfront_1_0', 'zipcode_98039', 'zipcode_98112', 'grade_13', 'grade_12', 'grade_11', 'grade_10', 'grade_6', 'zipcode_98040', 'grade_9', 'zipcode_98033', 'zipcode_98023', 'view_4_0', 'zipcode_98105', 'zipcode_98119', 'zipcode_98199', 'zipcode_98103', 'zipcode_98115', 'zipcode_98117', 'zipcode_98109', 'zipcode_98107', 'zipcode_98102', 'zipcode_98122', 'view_3_0', 'zipcode_98116', 'zipcode_98052', 'zipcode_98144', 'zipcode_98006', 'zipcode_98008', 'zipcode_98005', 'zipcode_98053', 'zipcode_98034', 'zipcode_98136', 'zipcode_98029', 'zipcode_98092', 'zipcode_98125', 'zipcode_98177', 'condition_5', 'zipcode_98007', 'zipcode_98126', 'zipcode_98074', 'zipcode_98075', 'zipcode_98027', 'zipcode_98133', 'zipcode_98118', 'view_2_0', 'zipcode_98072', 'zipcode_98155', 'sqft_above', 'view_1_0', 'yr_renovated', 'zipcode_98028', 'sqft_lot', 'zipcode_98106', 'zipcode_98011', 'condition_4', 'sqft_living15', 'bedrooms_3', 'zipcode_98108', 'floors_3_0', 'floors_2_0', 'bathrooms', 'zipcode_98146', 'zipcode_98056', 'zipcode_98077', 'zipcode_98024', 'zipcode_98065', 'zipcode_98059', 'zipcode_98045', 'zipcode_98019', 'bedrooms_2', 'zipcode_98014', 'zipcode_98166', 'bedrooms_7', 'zipcode_98168', 'floors_2_5', 'yr_built', 'bedrooms_9', 'zipcode_98038', 'zipcode_98055', 'zipcode_98058', 'bedrooms_10', 'zipcode_98010', 'zipcode_98148', 'sqft_lot15', 'zipcode_98178']

In [ ]:
len(b)

In [ ]:
print(a)

In [ ]:
len(x_cols)

In [ ]:
len(x_cols)

## Model Refinement I

In [ ]:
outcome = 'price'
x_cols = ['grade_7', 'sqft_living15', 'sqft_living', 'grade_6', 'zipcode_98023', 'zipcode_98004', 'grade_8', 'grade_5', 'grade_9', 'waterfront_1_0', 'zipcode_98003', 'zipcode_98092', 'zipcode_98038', 'zipcode_98002', 'zipcode_98032', 'zipcode_98039', 'bathrooms', 'zipcode_98033', 'zipcode_98040', 'zipcode_98112', 'zipcode_98115', 'zipcode_98117', 'zipcode_98103', 'zipcode_98199', 'zipcode_98105', 'zipcode_98119', 'zipcode_98107', 'zipcode_98116', 'zipcode_98109', 'zipcode_98122', 'zipcode_98102', 'view_4_0', 'zipcode_98006', 'zipcode_98052', 'zipcode_98136', 'zipcode_98144', 'sqft_lot', 'zipcode_98008', 'zipcode_98005', 'grade_4', 'zipcode_98053', 'view_3_0', 'zipcode_98029', 'zipcode_98126', 'zipcode_98125', 'zipcode_98034', 'zipcode_98177', 'zipcode_98007', 'condition_5', 'zipcode_98074', 'zipcode_98075', 'grade_10', 'zipcode_98027', 'view_2_0', 'sqft_above', 'zipcode_98133', 'zipcode_98118', 'zipcode_98072', 'zipcode_98155', 'condition_4', 'condition_3', 'view_1_0', 'zipcode_98011', 'zipcode_98065', 'zipcode_98028', 'zipcode_98077', 'zipcode_98106', 'yr_renovated', 'zipcode_98108', 'zipcode_98059', 'zipcode_98056', 'zipcode_98024', 'grade_11', 'floors_3_0', 'zipcode_98166', 'zipcode_98146', 'zipcode_98045', 'zipcode_98019', 'zipcode_98014', 'zipcode_98070', 'zipcode_98010', 'zipcode_98058', 'zipcode_98178', 'zipcode_98055', 'sqft_basement', 'sqft_lot15', 'bedrooms_7', 'zipcode_98148', 'grade_13', 'bedrooms_2', 'condition_2', 'floors_2_0', 'bedrooms_6', 'zipcode_98188', 'zipcode_98168', 'bedrooms_5', 'bedrooms_9']

In [ ]:
# Fitting the actual model
predictors = '+'.join(x_cols)
formula = outcome + '~' + predictors
model = ols(formula=formula, data=df_feats).fit()
model.summary()

In [ ]:
X = df_feats[x_cols]

In [ ]:
X.head()

In [ ]:
fig = sm.graphics.qqplot(model.resid, dist=stats.norm, line='45', fit=True)

In [ ]:
plt.scatter(model.predict(df_feats[x_cols]), model.resid)
plt.plot(model.predict(df_feats[x_cols]), [0 for i in range(len(df_feats))])

In [ ]:
vif = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
list(zip(x_cols, vif))

## Cross Validation

In [ ]:
linreg = LinearRegression()
cvs_10=cross_val_score(linreg, X, y, cv=10)
cvs_10_res_mean = round(np.mean(cvs_10),5)
cv_10_res_std = round(np.std(cvs_10),5)
print(f'Mean cross validation score for 10 fold is {cvs_10_res_mean}')
print(f'Standard deviation of the score is {cv_10_res_std}')


##  Location! Location! Location!

In [ ]:
# income_df=pd.read_csv('kc_house_data.csv')

In [ ]:
from uszipcode import SearchEngine, Zipcode


In [ ]:
search = SearchEngine(simple_zipcode=False)

In [ ]:
def zco(x):
    income = search.by_zipcode(x)
    
    return income.median_household_income if income else x
   

In [ ]:
income_df['income'] = income_df['zipcode'].astype(int).astype(str).apply(zco)

In [ ]:
income_df.head()

In [ ]:
# income_df=pd.read_csv('kc_house_data.csv')

In [ ]:
fig = plt.figure(figsize = (20,7))
fig = sns.boxplot(x='zipcode', y='price', data=income_df, showfliers = False)
plt.title('Prices of Houses Sold in King County by Zip Code')
plt.xlabel("Zip Code")
plt.xticks(rotation=90)
plt.ylabel('Sale Price, $')
plt.show()


https://www.zillow.com/bellevue-wa-98004/

https://www.zillow.com/mercer-island-wa-98040/

https://www.zillow.com/medina-wa-98039/

https://www.zillow.com/seattle-wa-98112/

In [ ]:
high_price_zipcodes=[98040,98004,98039,98112]
for i in high_price_zipcodes:
    zipcode = search.by_zipcode(i)
    income=zipcode.median_household_income
    print('Zip code:',i ,'-------   Median household income of all the houses in this zipcode:' ,income)